base/samples/Sonar.hpp

```diff
@@ -0,0 +1,289 @@
+    {
+        base::Angle angle;
+        bearings.resize(beam_count);
+        for (uint32_t i = 0; i < beam_count; ++i, angle += interval)
+            bearings[i] = angle;
+    }
+
+    /** Add data for one beam
+     */
+    void pushBeam(std::vector<float> const& bins, base::Angle bearing)
+    {
+        pushBeam(bins);
+        bearings.push_back(bearing);
+    }
+
+    /** Add data for one beam
+     */
+    void pushBeam(std::vector<float> const& bins)
+    {
+        if (!timestamps.empty())
+            throw std::invalid_argument("cannot call pushBeam(bins): the structure uses per-beam timestamps, use pushBeams(time, bins) instead");
+
+        pushBeamBins(bins);
+    }
+
+    /** Add data for one beam
+     */
+    void pushBeam(base::Time const& beam_time, std::vector<float> const& beam_bins, base::Angle bearing)
+    {
+        pushBeam(beam_time, beam_bins);
+        bearings.push_back(bearing);
+    }
+
+    /** Add data for one beam
+     */
+    void pushBeam(base::Time const& beam_time, std::vector<float> const& beam_bins)
+    {
+        pushBeamBins(beam_bins);
+        timestamps.push_back(beam_time);
+    }
+
+    /** Adds a set of bins to the bin set, updating beam_bins
+     *
+     * One usually does not use this method directly, but one of the overloaded
+     * pushBeam method
+     */
+    void pushBeamBins(std::vector<float> const& beam_bins)
+    {
+        if (beam_bins.size() != bin_count)
+            throw std::invalid_argument("pushBeam: the provided beam does not match the expected bin_count");
+        bins.insert(bins.end(), beam_bins.begin(), beam_bins.end());
+        beam_count++;
+    }
+
+    /** Returns the bearing of a given beam
+     *
+     * This is the bearing of the center of the beam. A zero bearing means the
+     * front of the device
+     */
+    base::Angle getBeamBearing(int beam) const
+    {
+        return bearings[beam];
+    }
+
+    /** Returns the bins of a given beam */
+    std::vector<float> getBeamBins(int beam) const
+    {
+        std::vector<float> bins;
+        getBeamBins(bins);
+        return bins;
+    }
+
+    /** Copies the bins of a given beam */
+    void getBeamBins(int beam, std::vector<float>& beam_bins) const
+    {
+        beam_bins.resize(bin_count);
+        std::copy(beam_bins.begin(), beam_bins.end(); &bins[beam * bin_count]);
+    }
+
+    /** Verify this structure's consistency
+     */
+    void validate()
+    {
+        if (bin_count * beam_count != bins.size())
+            throw std::logic_error("the number of elements in 'bins' does not match the bin and beam counts");
+        if (!timestamps.empty() && timestamps.size() != beam_count)
+            throw std::logic_error("the number of elements in 'timestamps' does not match the beam count");
+        if (bearings.size() != beam_count)
+            throw std::logic_error("the number of elements in 'bearings' does not match the beam count");
+    }
+};
```

Comment:
Maybe it would make sense to create a constructor/Setter from the old SonarScan type to make backward-compatibility easier...